### ISSUE TITLE
mount.s3ql crashes with "No Upload Threads available"

### ISSUE THREAD

**Issue by snatella** [s3ql/s3ql#234]
I'm getting a "Query produced 0 result rows" crashing `mount.s3ql` on any freshly created S3 s3ql filesystem, an example of what I'm doing and the error is below. Weirdly after the `mount.s3ql` crashes, and I `fusermount -u`, a `fsck.s3ql` will upload the data, but remounting and copying new data into the mount will again crash the `mount.s3ql` process with the same error again.

I've built s3ql 3.7.0 on CentOS7, with python3.8 provided by their SCL and most of the dependencies via pip (and they appear to be the versions as required on the install page). The tests excluding the rsync one work great (failing on lsetxattr - so not s3ql related) so I'm a bit stumped here.

```
$ mkfs.s3ql --log syslog s3://region/my-bucket/server/
Before using S3QL, make sure to read the user's guide, especially
the 'Important Rules to Avoid Losing Data' section.
Enter encryption password:
Confirm encryption password:
Generating random encryption key...
Creating metadata tables...
Dumping metadata...
Dumping metadata...
..objects..
..blocks..
..inodes..
..inode_blocks..
..symlink_targets..
..names..
..contents..
..ext_attributes..
Compressing and uploading metadata...
Wrote 162 bytes of compressed metadata.
Cycling metadata backups...
Backing up old metadata...
Please store the following master key in a safe location. It allows
decryption of the S3QL file system in case the storage objects holding
this information get corrupted:
---BEGIN MASTER KEY---
-snip-
---END MASTER KEY---
$ mount.s3ql --log syslog --allow-root s3://region/my-bucket/server/ /mnt/s3backup/
Using 8 upload threads.
Autodetected 4040 file descriptors available for cache entries
Using cached metadata.
Setting cache size to 35718 MB
Mounting s3://region/my-bucket/server/ at /mnt/s3backup...
$ tail /var/log/messages
Jan  4 17:50:37 localhost journal: mount.s3ql[4622:MainThread] s3ql.mount.determine_threads: Using 8 upload threads.
Jan  4 17:50:37 localhost journal: mount.s3ql[4622:MainThread] s3ql.mount.main: Autodetected 4040 file descriptors available for cache entries
Jan  4 17:50:39 localhost journal: mount.s3ql[4622:MainThread] s3ql.mount.get_metadata: Using cached metadata.
Jan  4 17:50:39 localhost journal: mount.s3ql[4622:MainThread] s3ql.mount.main_async: Setting cache size to 35718 MB
Jan  4 17:50:39 localhost journal: mount.s3ql[4622:MainThread] s3ql.block_cache.__init__: Loaded 0 entries from cache
Jan  4 17:50:39 localhost journal: mount.s3ql[4622:MainThread] s3ql.mount.main_async: Mounting s3://region/my-bucket/server/ at /mnt/s3backup...
Jan  4 17:50:39 localhost journal: mount.s3ql[4635:MainThread] s3ql.daemonize.detach_process_context: Daemonizing, new PID is 4636
$ cp testfile /mnt/s3backup/
$ tail /var/log/messages
Jan  4 17:34:11 localhost journal: mount.s3ql[3257:MainThread] s3ql.mount.unmount: Unmounting file system...
Jan  4 17:34:11 localhost journal: mount.s3ql[3257:MainThread] root.excepthook: Uncaught top-level exception:
Traceback (most recent call last):
  File "/opt/rh/rh-python38/root/usr/local/lib64/python3.8/site-packages/s3ql-3.7.0-py3.8-linux-x86_64.egg/s3ql/database.py", line 143, in get_row
    row = next(res)
StopIteration

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/rh/rh-python38/root/usr/local/lib64/python3.8/site-packages/s3ql-3.7.0-py3.8-linux-x86_64.egg/s3ql/block_cache.py", line 472, in upload_if_dirty
    block_id = self.db.get_val('SELECT id FROM blocks WHERE hash=?', (hash_,))
  File "/opt/rh/rh-python38/root/usr/local/lib64/python3.8/site-packages/s3ql-3.7.0-py3.8-linux-x86_64.egg/s3ql/database.py", line 127, in get_val
    return self.get_row(*a, **kw)[0]
  File "/opt/rh/rh-python38/root/usr/local/lib64/python3.8/site-packages/s3ql-3.7.0-py3.8-linux-x86_64.egg/s3ql/database.py", line 145, in get_row
    raise NoSuchRowError()
s3ql.database.NoSuchRowError: Query produced 0 result rows

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/rh/rh-python38/root/usr/local/bin/mount.s3ql", line 11, in <module>
    load_entry_point('s3ql==3.7.0', 'console_scripts', 'mount.s3ql')()
  File "/opt/rh/rh-python38/root/usr/local/lib64/python3.8/site-packages/s3ql-3.7.0-py3.8-linux-x86_64.egg/s3ql/mount.py", line 131, in main
    trio.run(main_async, options, stdout_log_handler)
  File "/opt/rh/rh-python38/root/usr/local/lib/python3.8/site-packages/trio/_core/_run.py", line 1928, in run
    raise runner.main_task_outcome.error
  File "/opt/rh/rh-python38/root/usr/local/lib64/python3.8/site-packages/s3ql-3.7.0-py3.8-linux-x86_64.egg/s3ql/mount.py", line 289, in main_async
    unmount_clean = True
  File "/opt/rh/rh-python38/root/usr/local/lib/python3.8/site-packages/trio/_core/_run.py", line 815, in __aexit__
    raise combined_error_from_nursery
  File "/opt/rh/rh-python38/root/usr/local/lib64/python3.8/site-packages/s3ql-3.7.0-py3.8-linux-x86_64.egg/s3ql/mount.py", line 749, in run
    await self.block_cache.upload_if_dirty(el)
  File "/opt/rh/rh-python38/root/usr/local/lib64/python3.8/site-packages/s3ql-3.7.0-py3.8-linux-x86_64.egg/s3ql/block_cache.py", line 491, in upload_if_dirty
    await self._queue_upload((el, obj_id))
  File "/opt/rh/rh-python38/root/usr/local/lib64/python3.8/site-packages/s3ql-3.7.0-py3.8-linux-x86_64.egg/s3ql/block_cache.py", line 536, in _queue_upload
    raise NoWorkerThreads('no upload threads')
s3ql.block_cache.NoWorkerThreads: no upload threads
```

Further interactions produce `OSError: [Errno 107] Transport endpoint is not connected: '/mnt/s3backup'`

```
$ fusermount -u /mnt/s3backup/
$ fsck.s3ql s3://region/my-bucket/server/
Starting fsck of s3://region/my-bucket/server/
Using cached metadata.
WARNING: Remote metadata is outdated.
Checking DB integrity...
Creating temporary extra indices...
Checking lost+found...
WARNING: Cleaning up interrupted upload of object 1
WARNING: Cleaning up interrupted upload of object 2
WARNING: Cleaning up interrupted upload of object 3
WARNING: Cleaning up interrupted upload of object 4
WARNING: Cleaning up interrupted upload of object 5
WARNING: Cleaning up interrupted upload of object 6
WARNING: Cleaning up interrupted upload of object 7
WARNING: Cleaning up interrupted upload of object 8
WARNING: Cleaning up interrupted upload of object 9
Checking for dirty cache objects...
..processed 1/17 files (5%)..WARNING: Writing dirty block 0 of inode 4 to backend
WARNING: Writing dirty block 2 of inode 5 to backend
..processed 3/17 files (17%)..WARNING: Writing dirty block 1 of inode 5 to backend
..processed 4/17 files (23%)..WARNING: Writing dirty block 0 of inode 5 to backend
..processed 5/17 files (29%)..WARNING: Writing dirty block 7 of inode 5 to backend
..processed 6/17 files (35%)..WARNING: Writing dirty block 12 of inode 5 to backend
..processed 7/17 files (41%)..WARNING: Writing dirty block 14 of inode 5 to backend
..processed 8/17 files (47%)..WARNING: Writing dirty block 15 of inode 5 to backend
WARNING: Writing dirty block 11 of inode 5 to backend
..processed 10/17 files (58%)..WARNING: Writing dirty block 8 of inode 5 to backend
..processed 11/17 files (64%)..WARNING: Writing dirty block 6 of inode 5 to backend
..processed 12/17 files (70%)..WARNING: Writing dirty block 3 of inode 5 to backend
..processed 13/17 files (76%)..WARNING: Writing dirty block 5 of inode 5 to backend
..processed 14/17 files (82%)..WARNING: Writing dirty block 10 of inode 5 to backend
..processed 15/17 files (88%)..WARNING: Writing dirty block 9 of inode 5 to backend
..processed 16/17 files (94%)..WARNING: Writing dirty block 4 of inode 5 to backend
..processed 17/17 files (100%)..WARNING: Writing dirty block 13 of inode 5 to backend
Checking names (refcounts)...
Checking contents (names)...
Checking contents (inodes)...
Checking contents (parent inodes)...
Checking objects (reference counts)...
Checking objects (backend)...
..processed 0 objects so far..
Checking objects (sizes)...
Checking blocks (referenced objects)...
Checking blocks (refcounts)...
WARNING: Block 1 has wrong reference count, setting from 0 to 1
WARNING: Block 2 has wrong reference count, setting from 0 to 1
WARNING: Block 3 has wrong reference count, setting from 0 to 1
WARNING: Block 5 has wrong reference count, setting from 0 to 1
WARNING: Block 6 has wrong reference count, setting from 0 to 1
WARNING: Block 7 has wrong reference count, setting from 0 to 1
WARNING: Block 8 has wrong reference count, setting from 0 to 1
Checking blocks (checksums)...
Checking inode-block mapping (blocks)...
Checking inode-block mapping (inodes)...
Checking inodes (refcounts)...
Checking inodes (sizes)...
Checking extended attributes (names)...
Checking extended attributes (inodes)...
Checking symlinks (inodes)...
Checking directory reachability...
Checking unix conventions...
Checking referential integrity...
Dropping temporary indices...
Dumping metadata...
..objects..
..blocks..
..inodes..
..inode_blocks..
..symlink_targets..
..names..
..contents..
..ext_attributes..
Compressing and uploading metadata...
Wrote 1.14 KiB of compressed metadata.
Cycling metadata backups...
Backing up old metadata...
Cleaning up local metadata...
Completed fsck of s3://region/my-bucket/server/
```

(but as mentioned, a remount and addition of new files will produce the same cycle of error log and crash again, which can be fusermount -u'ed and fsck'ed again etc)

**Comment by Nikratio** [comment]
Thanks for the report! The root cause here is the "no upload threads" message at the end, the exceptions before that do not signal a problem.

That said, I have no idea why the threads are not started (or maybe crash immediately?) in your case, when it seems to work for everyone else.

I'll try to give it a bit more thought, but sadly I do not have much time to spare on S3QL these days.

**Comment by angristan** [comment]
hello 👋 I have the same issue after upgrading from v3.3.2 (debian packaged) to 3.7.0 from the github release.

**Comment by angristan** [comment]
I downgraded to version 3.6.0, I didn't encounter the error (yet?). Maybe it was introduced in 3.7.0? Hope that helps.

**Comment by Nikratio** [comment]
Thanks for chiming in! Is there any way you could do a git bisect to identify the exact commit that introduced the problem?

**Comment by angristan** [comment]
I'm pretty sure it was introduced with https://github.com/s3ql/s3ql/commit/fbd3116780ccc6961422f9d0e78221055efe947e. I don't have the issue without this commit.

**Comment by Nikratio** [comment]
I'm stumped. The commit should only affect the logic on unmount, and not change anything about how mount.s3ql runs. I also continue to be unable to reproduce this.

Could you create a shell script that reproduces the problem? In particular, can you try to use the `local` backend and use something like `echo foobar > mnt/testfile` (to make sure we're writing the same data)? 

Finally, could you also try sleeping for a few seconds between mount.s3ql and writing the test file?

**Comment by angristan** [comment]
I was able to reproduce the bug with the `local` backend, with a big enough file.

```
root@sana ~/s3ql# git status
HEAD detached at release-3.7.0
nothing to commit, working tree clean
root@sana ~/s3ql# mkdir ~/fs
root@sana ~/s3ql# ./bin/mkfs.s3ql --plain local:///root/fs/
Before using S3QL, make sure to read the user's guide, especially
the 'Important Rules to Avoid Losing Data' section.
Creating metadata tables...
Dumping metadata...
Dumping metadata...
..objects..
..blocks..
..inodes..
..inode_blocks..
..symlink_targets..
..names..
..contents..
..ext_attributes..
Compressing and uploading metadata...
Wrote 451 bytes of compressed metadata.
Cycling metadata backups...
Backing up old metadata...
root@sana ~/s3ql# ./bin/mount.s3ql --metadata-upload-interval 60 local:///root/fs/ /mnt/
Using 10 upload threads.
Autodetected 1048514 file descriptors available for cache entries
Using cached metadata.
Setting cache size to 20314 MB
Mounting local:///root/fs/ at /mnt...
root@sana ~# echo foobar > /mnt/testfile
```

At first, nothing happened...

```
2021-01-20 15:15:18.709 22218:MainThread s3ql.mount.determine_threads: Using 10 upload threads.
2021-01-20 15:15:18.709 22218:MainThread s3ql.mount.main: Autodetected 1048514 file descriptors available for cache entries
2021-01-20 15:15:18.769 22218:MainThread s3ql.mount.get_metadata: Using cached metadata.
2021-01-20 15:15:18.790 22218:MainThread s3ql.mount.main_async: Setting cache size to 20314 MB
2021-01-20 15:15:18.795 22218:MainThread s3ql.mount.main_async: Mounting local:///root/fs/ at /mnt...
2021-01-20 15:15:18.802 22234:MainThread s3ql.daemonize.detach_process_context: Daemonizing, new PID is 22235
2021-01-20 15:16:18.824 22235:MainThread s3ql.mount.run: File system unchanged, not uploading metadata.
2021-01-20 15:17:18.831 22235:MainThread s3ql.mount.run: Dumping metadata...
2021-01-20 15:17:18.832 22235:MainThread s3ql.metadata.dump_metadata: ..objects..
2021-01-20 15:17:18.833 22235:MainThread s3ql.metadata.dump_metadata: ..blocks..
2021-01-20 15:17:18.834 22235:MainThread s3ql.metadata.dump_metadata: ..inodes..
2021-01-20 15:17:18.835 22235:MainThread s3ql.metadata.dump_metadata: ..inode_blocks..
2021-01-20 15:17:18.836 22235:MainThread s3ql.metadata.dump_metadata: ..symlink_targets..
2021-01-20 15:17:18.837 22235:MainThread s3ql.metadata.dump_metadata: ..names..
2021-01-20 15:17:18.838 22235:MainThread s3ql.metadata.dump_metadata: ..contents..
2021-01-20 15:17:18.838 22235:MainThread s3ql.metadata.dump_metadata: ..ext_attributes..
2021-01-20 15:17:18.840 22235:trio-worker-1 s3ql.metadata.upload_metadata: Compressing and uploading metadata...
2021-01-20 15:17:18.842 22235:trio-worker-1 s3ql.metadata.upload_metadata: Wrote 534 bytes of compressed metadata.
2021-01-20 15:17:18.842 22235:trio-worker-1 s3ql.metadata.upload_metadata: Cycling metadata backups...
2021-01-20 15:17:18.842 22235:trio-worker-1 s3ql.metadata.cycle_metadata: Backing up old metadata...
2021-01-20 15:18:18.859 22235:MainThread s3ql.mount.run: File system unchanged, not uploading metadata.
2021-01-20 15:19:18.865 22235:MainThread s3ql.mount.run: File system unchanged, not uploading metadata.
```

I tried with these files:

```
root@sana ~# ll /mnt/
total 169M
-rw-r--r-- 1 root root 141M Jan 20 16:17 archlinux-bootstrap-2021.01.01-x86_64.tar.gz
-rwxr-xr-x 1 root root  28M Jan 20 15:24 juicefs*
drwx------ 1 root root    0 Jan 20 15:20 lost+found/
-rw-r--r-- 1 root root 878K Jan 20 15:22 s3ql-3.7.0.tar.bz2
-rw-r--r-- 1 root root    7 Jan 20 15:21 testfile
```

I was able to trigger the exception consistently with `archlinux-bootstrap-2021.01.01-x86_64.tar.gz` only.

```
2021-01-20 16:22:01.412 36415:MainThread s3ql.mount.determine_threads: Using 10 upload threads.
2021-01-20 16:22:01.413 36415:MainThread s3ql.mount.main: Autodetected 1048514 file descriptors available for cache entries
2021-01-20 16:22:01.473 36415:MainThread s3ql.mount.get_metadata: Using cached metadata.
2021-01-20 16:22:01.481 36415:MainThread s3ql.mount.main_async: Setting cache size to 20147 MB
2021-01-20 16:22:01.485 36415:MainThread s3ql.mount.main_async: Mounting local:///root/fs/ at /mnt...
2021-01-20 16:22:01.490 36431:MainThread s3ql.daemonize.detach_process_context: Daemonizing, new PID is 36432
2021-01-20 16:22:26.734 36432:MainThread s3ql.mount.unmount: Unmounting file system...
2021-01-20 16:22:26.745 36432:MainThread root.excepthook: Uncaught top-level exception:
Traceback (most recent call last):
  File "/root/s3ql/src/s3ql/database.py", line 143, in get_row
    row = next(res)
StopIteration

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/root/s3ql/src/s3ql/block_cache.py", line 472, in upload_if_dirty
    block_id = self.db.get_val('SELECT id FROM blocks WHERE hash=?', (hash_,))
  File "/root/s3ql/src/s3ql/database.py", line 127, in get_val
    return self.get_row(*a, **kw)[0]
  File "/root/s3ql/src/s3ql/database.py", line 145, in get_row
    raise NoSuchRowError()
s3ql.database.NoSuchRowError: Query produced 0 result rows

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./bin/mount.s3ql", line 21, in <module>
  File "/root/s3ql/src/s3ql/mount.py", line 131, in main
    trio.run(main_async, options, stdout_log_handler)
  File "/usr/local/lib/python3.7/dist-packages/trio/_core/_run.py", line 1795, in run
    raise runner.main_task_outcome.error
  File "/root/s3ql/src/s3ql/mount.py", line 289, in main_async
    unmount_clean = True
  File "/usr/local/lib/python3.7/dist-packages/trio/_core/_run.py", line 724, in __aexit__
    raise combined_error_from_nursery
  File "/root/s3ql/src/s3ql/mount.py", line 749, in run
    await self.block_cache.upload_if_dirty(el)
  File "/root/s3ql/src/s3ql/block_cache.py", line 491, in upload_if_dirty
    await self._queue_upload((el, obj_id))
  File "/root/s3ql/src/s3ql/block_cache.py", line 536, in _queue_upload
    raise NoWorkerThreads('no upload threads')
s3ql.block_cache.NoWorkerThreads: no upload threads
```

Maybe it's because once the file is big enough, s3ql uses multiple upload threads simultaneously? 

Hope that helps.

**Comment by Nikratio** [comment]
Thanks! Could you also try if it matters if you do the copy right after mounting, or if you wait a minute and then copy?

**Comment by angristan** [comment]
It doesn't seem to matter 🙂

**Comment by r0ps3c** [comment]
I think this is a block cache initialisation issue introduced in fbd3116. Should be fixed by #240

**Comment by angristan** [comment]
Can confirm #240 fixes the issue. Thanks @r0ps3c 🙌
